File: go/pe-004.go

package main

import (
    "fmt"
    "helpers"
)

// Three digits isn't much. Let's brute force this one.

// NOTE -- There are certainly ways we could have done this prettier.
// Onward and upward! 

// =====================================================================

func is_palindrome(n int) bool {
    digits := []int{}
    for n>0 {
        digits = append(digits, n % 10)
        n = n/10
    }
    for i, dig := range digits {
        if dig != digits[len(digits)-i-1] { return false }
    }
    return true
}

// =====================================================================

func main() {
    palindromes := []int{}
    for i:=100; i<1000; i++ {
        for j:=i; j<1000; j++ {
            if is_palindrome(i*j) { palindromes = append(palindromes, i*j) }
        }
    }
    fmt.Println( helpers.Max(palindromes) )
}

File: python/pe-004.py
#!/usr/bin/env python3

"""
Charles McEachern
2018-04-17
https://projecteuler.net/problem=4
"""

# ######################################################################

def main():
    # Three digits isn't very dig. Let's brute force this one. 
    print( max( find_palindrome_numbers(3) ) )

# ----------------------------------------------------------------------

def find_palindrome_numbers(ndigs=3):
    imin = 10**(ndigs-1)
    imax = 10**ndigs
    for i in range(imin, imax):
        for j in range(imin, imax):
            if is_palindrome_number(i*j):
                yield i*j

# ----------------------------------------------------------------------

def is_palindrome_number(n):
    return str(n) == str(n)[::-1]

# ######################################################################

if __name__ == '__main__':
    main()
